File: [구현] 비밀지도.cpp
#include <string>
#include <vector>
#include <bitset>
#include <iostream>
#include <algorithm>
using namespace std;

#define BIT_SIZE 16 /* (1 ≦ N ≦ 16) */
#define INT_PAIR pair<int, int>

const string convertBinaryDigit(const INT_PAIR value, const int n) {

	bitset<BIT_SIZE> binary(value.first);
	binary |= bitset<BIT_SIZE>(value.second);

	const string answer = binary.to_string(' ', '#');
	return string(answer.rbegin(), answer.rbegin() + n);
}

// 원래의 비밀지도를 해독하여 '#', 공백으로 구성된 문자열 배열로 출력하라.
vector<string> solution(int n, vector<int> arr1, vector<int> arr2) {

	vector<string> answer;
	
	for (int index = 0; index < arr1.size(); index++) {
		const INT_PAIR digit = make_pair(arr1[index], arr2[index]);

		string map = convertBinaryDigit(digit, n);
		std::reverse(map.begin(), map.end());

		answer.push_back(map);
	}

	return answer;
}

File: [구현] 비밀지도.swift
/**
    - NOTE:
        - 지도는 한 변의 길이가 n인 정사각형 배열 형태로, 각 칸은 "공백"(" ") 또는 "벽"("#") 두 종류로 이루어져 있다.
        - 전체 지도는 두 장의 지도를 겹쳐서 얻을 수 있다. 각각 "지도 1"과 "지도 2"라고 하자. 지도   1 또는 지도 2 중 어느 하나라도 벽인 부분은 전체 지도에서도 벽이다. 지도 1과 지도 2에서 모두 공백인 부분은 전체 지도에서도 공백이다.
        - "지도 1"과 "지도 2"는 각각 정수 배열로 암호화되어 있다.
        - 암호화된 배열은 지도의 각 가로줄에서 벽 부분을 1, 공백 부분을 0으로 부호화했을 때 얻어지는 이진수에 해당하는 값의 배열이다.
*/
func solution(_ n: Int, _ arr1: [Int], _ arr2: [Int]) -> [String] {
    
    var answer: [String] = []
    
    let length = arr1.count
    for index in Int.zero..<length {
        let leftValue = arr1[index]
        let rightValue = arr2[index]
        
        var result = String(leftValue | rightValue, radix: 2)
        
        // 계산되어 진 값에 대한 길이가 동일하지 않은 경우
        while result.count != n {
            result.insert("0", at: result.startIndex)
        }
        
        result = result.compactMap { $0 == "0" ? " " : "#" }.joined()
        answer.append(result)
    }
    
    return answer
}
